### DISCUSSION TITLE
Images as a Terminal background feature

### DISCUSSION THREAD

**Discussion by ghost** [termux/termux-app#2250] [Ideas]
Any statuses for
https://github.com/termux/termux-app/pull/1918

I'd open a discussion since it would be nice to add such customizations

although there could be any disadvantages for this one if added 🤔

**Comment by DethByte64** [comment]
Or transparent so we could just use our wallpaper

**Comment by agnostic-apollo** [comment]
I have no issue with something like that being added as a feature since I know lot of users want aesthetically pleasing stuff. However, I'll not do it myself, at least currently, I don't have time. Pull request need to be sent/updated.

Moreover, I personally will not merge a messy code like the one in that pull request in its current state. I have spent countless hours fixing old termux code and making it more readable and compartmentalized, don't expect me to approve messy code again and make it return to its old state.

- Things need to put into dedicated classes and sub packages, `TermuxActivity` has no business handling `SharedPreferences`, file handling, file choosers, etc. It should only have minimal code added that calls other classes. Further unnecessary stuff should not be added to `com.termux.app` package but to appropriate sub package.
- The `View` for background image and overlay should be `disabled` by default, instead of just `null`. Large photos should be compressed and also as per device display resolution, imagine a user selecting an image that is `2-10MB` or more, that will/may be detrimental to performance depending on device and will likely cause an `OOM` and crash termux, depending on device state. There is currently some compression code, but output file size is not being checked. Size checks should be made, both on saving and loading, since user may modify file in background. Note that the image+all the terminal output buffers need to be held in memory. So if an image uses lots of memory, it will affect other parts of termux. For this, just a background color/tint would be best, it can be transparent too. The `TerminalEmulator` already has a background color, which is [set by termux](https://github.com/termux/termux-app/blob/v0.117/app/src/main/java/com/termux/app/terminal/TermuxTerminalSessionClient.java#L440) for the style, which alone might be enough for overlay tint or required for transparency (I haven't looked much at that code). 
- The image chooser button is best added to the `Style` option in context menu instead of drawer.
- Exceptions, errors and info should be logged appropriately with `Logger` class at appropriate log level and `LOG_TAG`.
- `FilePathFromDevice` should not have kitkat code check.
- `FileUtils` should be used for basic copying, deletion and stuff.
- Strings shown to user must be added to `strings.xml` with `msg_`, etc prefixes.
- The `"terminal_background.jpg"` should not be hardcoded. It's also best added as a constant in `TermuxContants` and the image stored in `~/.termux`. The `overlay` color should be loaded from `termux.properties` so that it can also be restored when user restores a backup, along with the image. There is no specific need for adding overlay color chooser, plenty of external sites for color codes, hex value validation should obviously be done. Also termux doesn't currently support saving values to `termux.properties`, so chooser isn't that useful.
- An option should be added to Termux Settings to disable background image loading, in case some faulty/malicious image causes termux to fail to start. Think "script kiddie" or just a bad image.

Once a new proof of concept is provided it can be looked into on how well it works and if any other changes need to be made.

**Comment by RalfWerner** [comment]
> Or transparent so we could just use our wallpaper

![grafik](https://user-images.githubusercontent.com/45426380/152645209-140891d5-441f-4fdb-8559-eac93992a31e.png)
Like this shot @DethByte64 (`*.jpg ` in background)? Here **Termux:float** is used over **fotos**-App. The same works also with **RealVNC** in parallel or **X11**, like in following screenshot (here with log output of XClient):

![grafik](https://user-images.githubusercontent.com/45426380/152645437-edbca71b-fbfd-413b-bee3-a60d82704392.png)
This is already possible now and after _merge_ of **app+float** and with a _common_ Toolbox (ExtraKeys) it would be perfect :)

**Comment by Anonymous2716** [comment]
On termux screen when showing colored stuff for example cmatrix in colored mode or show image in terminal it has issues like can't scroll without lagging or the fps decreases . I don't know what's the actual issue.  Did you all noticed the same?

**Reply by RalfWerner** [reply]
> can't scroll without lagging or the fps decreases . I don't know what's the actual issue. Did you all noticed the same?

Never noticed scroll lagging  or the fps decreases in **float**! Insert an example to check.
